File: frontend/src/features/production/domain/productionWorkbench.ts
import type { ProcedureOption } from './types'

export type WorkbenchActivity = {
  open_work_order_count: number
  processing_work_order_count: number
  ready_for_result_work_order_count: number
  pending_qc_work_order_count: number
  rework_work_order_count: number
}

export type StandardWorkbenchSource = {
  repository_id: number
  production_item_id: number
  source_work_order_id: number | null
  on_hand_quantity: number
  reserved_quantity: number
  available_quantity: number
  arrived_at: string | null
  available_procedures: ProcedureOption[]
  procedure_configuration_confirmed: boolean
  can_create_work_order: boolean
}

export type WorkbenchInventorySource = {
  repository_id: number
  production_item_id: number
  source_work_order_id: number | null
  on_hand_quantity: number
  reserved_quantity: number
  available_quantity: number
  arrived_at: string | null
}

export type AssemblyWorkbenchInputMaterial = {
  material_key: string
  item_code: string
  item_name: string
  unit_quantity: number
  on_hand_quantity: number
  reserved_quantity: number
  available_quantity: number
  allocated_quantity: number
  sources: WorkbenchInventorySource[]
}

export type AssemblyWorkbenchContinuationSource = WorkbenchInventorySource & {
  available_procedures: ProcedureOption[]
  procedure_configuration_confirmed: boolean
  can_create_work_order: boolean
}

export type StandardWorkbenchPosition = {
  position_key: string
  position_type: 'standard'
  production_item_id: number
  customer_order_item_id: number
  customer_order_no: string
  customer_name: string
  product_id: number
  product_version: number
  factory_code: string
  product_name: string
  item_code: string
  item_name: string
  flow_node_id: string
  source_flow_node_id: string
  source_node_label: string
  node_type: string
  workshop_id: number
  workshop_name: string
  department_id: number
  department_name: string
  department_code: string
  delivery_date: string
  arrived_at: string | null
  last_activity_at: string | null
  activity: WorkbenchActivity
  can_create_work_order: boolean
  on_hand_quantity: number
  reserved_quantity: number
  available_quantity: number
  source_count: number
  sources: StandardWorkbenchSource[]
}

export type AssemblyWorkbenchPosition = {
  position_key: string
  position_type: 'assembly'
  production_item_id: number | null
  customer_order_item_id: number
  customer_order_no: string
  customer_name: string
  product_id: number
  product_version: number
  factory_code: string
  product_name: string
  item_code: string
  item_name: string
  flow_node_id: string
  source_flow_node_id: null
  source_node_label: null
  node_type: string
  workshop_id: number
  workshop_name: string
  department_id: number
  department_name: string
  department_code: string
  delivery_date: string
  arrived_at: string | null
  last_activity_at: string | null
  activity: WorkbenchActivity
  can_create_work_order: boolean
  capacity_quantity: number
  initial_capacity_quantity: number
  continuation_capacity_quantity: number
  input_materials_complete: boolean
  can_create_initial_work_order: boolean
  input_material_count: number
  input_materials: AssemblyWorkbenchInputMaterial[]
  continuation_sources: AssemblyWorkbenchContinuationSource[]
  available_procedures: ProcedureOption[]
  procedure_configuration_confirmed: boolean
}

export type ProductionWorkbenchPosition =
  | StandardWorkbenchPosition
  | AssemblyWorkbenchPosition

export type ProductionWorkbenchPositionQuery = {
  page: number
  page_size: number
  customer_order_item_id: number
  workshop_id: number
  flow_node_id: string
  production_item_id?: number
}

File: backend/src/schemas/production_workbench.py
from datetime import date
from typing import Annotated, Literal

from pydantic import Field

from schemas.common import FlowNodeType
from schemas.organization import ProcedureResponse
from schemas.production_base import ProductionModel


class WorkbenchActivityResponse(ProductionModel):
    open_work_order_count: int = Field(ge=0)
    processing_work_order_count: int = Field(ge=0)
    ready_for_result_work_order_count: int = Field(ge=0)
    pending_qc_work_order_count: int = Field(ge=0)
    rework_work_order_count: int = Field(ge=0)


class WorkbenchInventorySourceResponse(ProductionModel):
    repository_id: int
    production_item_id: int
    source_work_order_id: int | None
    on_hand_quantity: int = Field(gt=0)
    reserved_quantity: int = Field(ge=0)
    available_quantity: int = Field(ge=0)
    arrived_at: str | None


class StandardWorkbenchSourceResponse(WorkbenchInventorySourceResponse):
    available_procedures: list[ProcedureResponse] = Field(default_factory=list)
    procedure_configuration_confirmed: bool
    can_create_work_order: bool


class WorkbenchPositionBase(ProductionModel):
    position_key: str
    position_type: str
    production_item_id: int | None
    customer_order_item_id: int
    customer_order_no: str
    customer_name: str
    product_id: int
    product_version: int
    factory_code: str
    product_name: str
    item_code: str
    item_name: str
    flow_node_id: str
    source_flow_node_id: str | None
    source_node_label: str | None
    node_type: FlowNodeType
    workshop_id: int
    workshop_name: str
    department_id: int
    department_name: str
    department_code: str
    delivery_date: date
    arrived_at: str | None
    last_activity_at: str | None
    activity: WorkbenchActivityResponse
    can_create_work_order: bool


class StandardWorkbenchPositionResponse(WorkbenchPositionBase):
    position_type: Literal["standard"]
    production_item_id: int
    source_flow_node_id: str
    source_node_label: str
    on_hand_quantity: int = Field(ge=0)
    reserved_quantity: int = Field(ge=0)
    available_quantity: int = Field(ge=0)
    source_count: int = Field(ge=0)
    sources: list[StandardWorkbenchSourceResponse] = Field(default_factory=list)


class AssemblyWorkbenchInputMaterialResponse(ProductionModel):
    material_key: str
    item_code: str
    item_name: str
    unit_quantity: int = Field(gt=0)
    on_hand_quantity: int = Field(ge=0)
    reserved_quantity: int = Field(ge=0)
    available_quantity: int = Field(ge=0)
    allocated_quantity: int = Field(ge=0)
    sources: list[WorkbenchInventorySourceResponse] = Field(default_factory=list)


class AssemblyWorkbenchContinuationSourceResponse(WorkbenchInventorySourceResponse):
    available_procedures: list[ProcedureResponse] = Field(default_factory=list)
    procedure_configuration_confirmed: bool
    can_create_work_order: bool


class AssemblyWorkbenchPositionResponse(WorkbenchPositionBase):
    position_type: Literal["assembly"]
    source_flow_node_id: None = None
    source_node_label: None = None
    capacity_quantity: int = Field(ge=0)
    initial_capacity_quantity: int = Field(ge=0)
    continuation_capacity_quantity: int = Field(ge=0)
    input_materials_complete: bool
    can_create_initial_work_order: bool
    input_material_count: int = Field(ge=0)
    input_materials: list[AssemblyWorkbenchInputMaterialResponse] = Field(
        default_factory=list,
    )
    continuation_sources: list[AssemblyWorkbenchContinuationSourceResponse] = Field(
        default_factory=list,
    )
    available_procedures: list[ProcedureResponse] = Field(default_factory=list)
    procedure_configuration_confirmed: bool


ProductionWorkbenchPositionResponse = Annotated[
    StandardWorkbenchPositionResponse | AssemblyWorkbenchPositionResponse,
    Field(discriminator="position_type"),
]


class ProductionWorkbenchPositionListEnvelope(ProductionModel):
    data: list[ProductionWorkbenchPositionResponse]
    total: int = Field(ge=0)


__all__ = [
    "AssemblyWorkbenchInputMaterialResponse",
    "AssemblyWorkbenchContinuationSourceResponse",
    "AssemblyWorkbenchPositionResponse",
    "ProductionWorkbenchPositionListEnvelope",
    "ProductionWorkbenchPositionResponse",
    "StandardWorkbenchPositionResponse",
    "StandardWorkbenchSourceResponse",
    "WorkbenchActivityResponse",
    "WorkbenchInventorySourceResponse",
]
